File: src/powderkeg_example/sql.clj
(ns powderkeg-example.sql
  (:require
   [powderkeg.core :as keg]
   [net.cgrand.xforms :as x])
  (:import
   [org.apache.spark.sql SparkSession SQLContext]
   [org.apache.spark.sql.types StringType StructField StructType]
   [org.apache.spark.sql.types DataTypes]
   [org.apache.spark.sql Row SaveMode RowFactory] ))

;; "local[*]" 连接本地的spark
(keg/connect! "local")

(def customer
  (let [;; 读取某个目录的某些文件*.csv: 返回JavaRDD
        txtf (.textFile keg/*sc* "/Users/clojure/Datas/2000W/100-test.csv")
        ;; Powderkeg: 实时，不用aot，原生transducer处理数据, (keg/rdd (map ...) (filter ...) (mapcat ...)). keg/rdd是transducer, 就像"->"一样使用
        ;; 测试或者转为Clojure输出: 要(into [] (keg/rdd ...))转为Clojure处理
        maped-rdd (keg/rdd
                   txtf
                   (map #(clojure.string/split % #",") )
                   (filter #(> (count %) 7) )
                   (map
                    ;; sql就是每一行都是row对象, Scala里面一般会建立一个类设置多个属性来对应它, Clojure 版本里面要转成Row才行(用defrecord无效)
                    #(RowFactory/create
                      ;; 转换为Java的数组
                      (into-array
                       ;; 取出第 0 5 4 6 7 列的数据
                       (mapv % [0 5 4 6 7])
                       )))
                   (distinct))]
    ;; 要把 Java的RDD变成 dataframe , 就要用 createDataFrame
    ;; 要传rdd及Schema(StructField)
    (.createDataFrame
     ;; keg/*sc* 就是Java的Spark源码的 sc
     ;; 从javaSparkSession得到scala版本的SparkSession
     (->> keg/*sc* .sc (new SparkSession))
     maped-rdd
     ;; 用StructField来创建StructType
     (DataTypes/createStructType
      (map #(DataTypes/createStructField % DataTypes/StringType false)
           ;;多个平行的字段
           ["name" "gender" "ctfId" "birthday" "address"]) )
     )
    )
  )

(comment
  ;; 注册table表格
  (.createOrReplaceTempView customer "customer")
  ;; 最终show出来,打印在repl里面,不会再Emacs返回
  (.show
   ;; 用dataframe和sql是差不多的, sql基于dataframe
   (.sql
    ;; SparkContext是基于rdd, SparkSession是基于DataFrame的
    ;; 从javaSparkContext得到scala版本的SparkContext
    (->> keg/*sc* .sc (new SQLContext))
    "SELECT * FROM customer LIMIT 10"
    )
   )
  )

File: src/scala/powderkeg_example/sql.scala
//Spark 2.1.1的版本: http://colobu.com/2014/12/11/spark-sql-quick-start/
val sqlContext = new org.apache.spark.sql.SQLContext(sc)

case class Customer(name: String, gender: String, ctfId: String, birthday: String, address: String)
val customer = sc.textFile("/Users/clojure/Datas/2000W/*.csv").map(_.split(",")).filter(line => line.length > 7).map(p => Customer(p(0), p(5), p(4), p(6), p(7))).distinct().toDF()
customer.registerTempTable("customer")

def toInt(s: String):Int = {
  try {
    s.toInt
  } catch {
    case e:Exception => 9999
  }
}

def myfun(birthday: String) : String = {
  var rt = "未知"
  if (birthday.length == 8) {
    val md = toInt(birthday.substring(4))
    if (md >= 120 & md <= 219)
      rt = "水瓶座"
    else if (md >= 220 & md <= 320)
      rt = "双鱼座"
    else if (md >= 321 & md <= 420)
      rt = "白羊座"
    else if (md >= 421 & md <= 521)
      rt = "金牛座"
    else if (md >= 522 & md <= 621)
      rt = "双子座"
    else if (md >= 622 & md <= 722)
      rt = "巨蟹座"
    else if (md >= 723 & md <= 823)
      rt = "狮子座"
    else if (md >= 824 & md <= 923)
      rt = "处女座"
    else if (md >= 924 & md <= 1023)
      rt = "天秤座"
    else if (md >= 1024 & md <= 1122)
      rt = "天蝎座"
    else if (md >= 1123 & md <= 1222)
      rt = "射手座"
    else if ((md >= 1223 & md <= 1231) | (md >= 101 & md <= 119))
      rt = "摩蝎座"
    else
      rt = "未知"
  }
  rt
}

sqlContext.udf.register("constellation",  (x:String) => myfun(x))
var result = sqlContext.sql("SELECT constellation(birthday), count(constellation(birthday)) FROM customer group by constellation(birthday)")
result.collect().foreach(println)

result = sqlContext.sql("SELECT gender, count(gender) FROM customer where gender = 'F' or gender = 'M' group by gender")
result.collect().foreach(println)
